### PR TITLE
amqp_client: forward errors to the calling thread

### PR DESCRIPTION
**Pull request by tehasdf**
Implement a `publish` method on the `AMQPConnection` that makes
sure errors while publishing are bubbled up, and use that method
in handlers.

### CODE REVIEW

**Comment by tehasdf** [comment]
You might think this is shit, but [no, it's necessary](https://www.youtube.com/watch?v=6ixvpLCdqkA)

File: cloudify/amqp_client.py [hunk 1]
```diff
@@ -263,7 +276,46 @@ def close(self, wait=True):
...
+        :param message: Kwargs for the pika basic_publish call. Should at
+                        least contain the "body" and "exchange" keys, and
+                        it might contain other keys such as "routing_key"
+                        or "properties"
+        :param wait: Whether to wait for the message to actually be sent.
+                     If true, an exception will be raised if the message
+                     cannot be sent.
+        """
+        if wait and self._consumer_thread \
+                and self._consumer_thread is threading.current_thread():
```
Comment (mcouthon) [inline cloudify/amqp_client.py]: If this behavior is broken, why is it allowed in the first place?

File: cloudify/amqp_client.py [hunk 2]
```diff
@@ -263,7 +276,46 @@ def close(self, wait=True):
...
+            # when sending from the connection thread, we can't wait because
+            # then we wouldn't allow the actual send loop (._process_publish)
+            # to run, because we'd block on the err_queue here
+            raise RuntimeError(
+                'Cannot wait when sending from the connection thread')
+
+        # the message will likely be sent from another thread (the .consume
```
Comment (mcouthon) [inline cloudify/amqp_client.py]: What do you mean "likely"? Isn't it always the case?

File: cloudify/amqp_client.py [hunk 3]
```diff
@@ -263,7 +276,46 @@ def close(self, wait=True):
...
+        # thread). If an error happens there, we must have a way to get it
+        # back out, so we pass a Queue together with the message, that will
+        # contain either an exception instance, or None
+        err_queue = Queue.Queue() if wait else None
+        envelope = {
+            'message': message,
+            'err_queue': err_queue
+        }
+        self._publish_queue.put(envelope)
+        err = err_queue.get()
```
Comment (mcouthon) [inline cloudify/amqp_client.py]: This will throw if `err_queue` is `None`

Comment (mcouthon) [inline cloudify/amqp_client.py]: Also, do we want it to block forever, or do we want a timeout here? Can there be an unhandled scenario on the other side, that will cause this thread to block forever?

File: cloudify/amqp_client.py [hunk 4]
```diff
@@ -263,7 +276,46 @@ def close(self, wait=True):
...
+        if isinstance(err, Exception):
+            raise err
```
Comment (mcouthon) [inline cloudify/amqp_client.py]: Do we want to handle the case where `err` is something other than an Exception?

File: cloudify/amqp_client.py (hunk 1 above)
Comment (tehasdf) [reply cloudify/amqp_client.py]: it's not, this line disallows it :P
other than that, we have no way to stop callbacks from calling `.publish`

File: cloudify/amqp_client.py (hunk 4 above)
Comment (tehasdf) [reply cloudify/amqp_client.py]: what else could it be? currently it's always None in the success case

File: cloudify/amqp_client.py [hunk 5]
```diff
@@ -263,7 +276,47 @@ def close(self, wait=True):
...
+        # back out, so we pass a Queue together with the message, that will
+        # contain either an exception instance, or None
+        err_queue = Queue.Queue() if wait else None
+        envelope = {
+            'message': message,
+            'err_queue': err_queue
+        }
+        self._publish_queue.put(envelope)
+        if err_queue:
+            err = err_queue.get(timeout=timeout)
```
Comment (tehasdf) [inline cloudify/amqp_client.py]: and yeah, about the timeout, nothing uses it atm, but perhaps some thing actually should...

**Pull request merged**